src/messages.js

```diff
@@ -1,143 +1,161 @@
+    if (Buffer.isBuffer(value.data)) {
+      if (offset + encode.bytes + value.data.length > buffer.length) {
+        throw new RangeError('destination buffer is too small')
+      }
+      value.data.copy(buffer, offset + encode.bytes)
+      encode.bytes += value.data.length
     }
-    return bytes
-  },
+    return buffer
+  }
 
-  function decode (buf, d) {
-    d.bytes = message.length
-    const value = rejectStruct.decode(buf)
-    if (message.length > rejectStruct.length) {
-      value.data = buf.slice(d.bytes, d.bytes + message.length)
+  function decode (buffer, offset, end) {
+    if (!offset) offset = 0
+    if (!end) end = buffer.length
+    var value = baseStruct.decode(buffer, offset, end)
+    decode.bytes = baseStruct.decode.bytes
+    if (decode.bytes === end) {
+      value.data = new Buffer(0)
     } else {
-      value.data = NULL_BUFFER
+      value.data = buffer.slice(decode.bytes, end)
+      decode.bytes = end
     }
     return value
-  },
+  }
 
   function encodingLength (value) {
-    return rejectStruct.length + (value.data ? value.data.length : 0)
+    var dataLength = Buffer.isBuffer(value.data) ? value.data.length : 0
+    return baseStruct.encodingLength(value) + dataLength
   }
-)
-const reject = exports.reject()
-exports.reject.encode = reject.encode
-exports.reject.encodingLength = reject.encodingLength
 
-exports.filterload = struct([
-  { name: 'data', type: struct.VarBuffer(varint) },
-  { name: 'nHashFuncs', type: struct.UInt32LE },
-  { name: 'nTweak', type: struct.UInt32LE },
-  { name: 'nFlags', type: struct.UInt8 }
-])
-
-exports.filteradd = struct([
-  { name: 'data', type: struct.VarBuffer(varint) }
-])
+  return { encode: encode, decode: decode, encodingLength: encodingLength }
+})()
 
-exports.merkleblock = struct([
-  { name: 'header', type: struct([
+// https://bitcoin.org/en/developer-reference#p2p-network
+// TODO: move to own files
+module.exports = {
+  // Data Messages
+  block: struct([
     { name: 'version', type: struct.Int32LE },
-    { name: 'prevHash', type: buffer32 },
-    { name: 'merkleRoot', type: buffer32 },
+    { name: 'prevHash', type: struct.Buffer(32) },
+    { name: 'merkleRoot', type: struct.Buffer(32) },
     { name: 'timestamp', type: struct.UInt32LE },
     { name: 'bits', type: struct.UInt32LE },
-    { name: 'nonce', type: struct.UInt32LE }
-  ]) },
-  { name: 'hashes', type: types.vararray(varint, buffer32) },
-  { name: 'flags', type: struct.VarBuffer(varint) }
-])
-
-exports.alert = struct([
-  { name: 'payload', type: struct.VarBuffer(varint) },
-  { name: 'signature', type: struct.VarBuffer(varint) }
-])
+    { name: 'nonce', type: struct.UInt32LE },
+    { name: 'transactions', type: struct.VarArray(varint, tx) }
+  ]),
+  getblocks: struct([
+    { name: 'version', type: struct.UInt32BE },
+    { name: 'locator', type: struct.VarArray(varint, struct.Buffer(32)) },
+    { name: 'hashStop', type: struct.Buffer(32) }
```

Comment:
Why are fixed-size Buffer types now all separate function calls? Are the types now memoized in `varstruct`? If not this uses slightly more memory than it needs to since each call to `struct.Buffer(x)` returns a new codec.